File: 2016/14/part1.rb
require 'digest'

salt = 'qzyelonm'
i = 0
counter = 0
hashes = Hash.new { |h, k| h[k] = Digest::MD5.hexdigest("#{salt}#{k}") }

while counter < 64
  if hashes[i].match /(\w)\1\1/
    char = $1
    j = i + 1
    while j < i + 1000
      if hashes[j].match /#{char * 5}/
        counter += 1
        if counter == 64
          puts i
          break
        end
      end
      j += 1
    end
  end
  i += 1
end

File: 2015/04/part1.ex
defmodule Crypto do
  def md5(data) do
    Base.encode16(:erlang.md5(data), case: :lower)
  end
end

defmodule AOC2015.Day04 do

  def solve(key) do
    key |> find_hash
  end

  defp find_hash(key, counter \\ 7, index \\ 0) do
    hash = Crypto.md5 "#{key}#{index}"
    if String.starts_with?(hash, "00000") do
        if counter == 0 do
          index
        else
          find_hash(key, counter - 1, index)
        end
    else
      find_hash(key, counter, index + 1)
    end
  end

end


AOC2015.Day04.solve("bgvyzdsv") |> IO.puts